src/ClaimToken.sol

```diff
@@ -0,0 +1,205 @@
+
+    // -----------------------
+    // -- Private Variables --
+    // -----------------------
+
+    mapping(address signer => bool isActivated) private _isActivatedSigner;
+    mapping(address token => mapping(bytes32 eventIDHash => mapping(address user => uint256 amount))) private
+        _userClaimedAmount;
+    mapping(address token => mapping(bytes32 eventIDHash => bool isEventOngoing)) private _isEventOngoing;
+    mapping(address token => mapping(bytes32 eventIDHash => bool isEventCreated)) private _isEventCreated;
+
+    // List of signers
+    EnumerableSet.AddressSet private _signerSet;
+
+    // -----------------
+    // -- Constructor --
+    // -----------------
+
+    constructor(address _admin, address[] memory signers) Ownable(_admin) {
+        bool[] memory isActivatedList = new bool[](signers.length);
+        for (uint256 i = 0; i < signers.length; i++) {
+            isActivatedList[i] = true;
+        }
+
+        _updateSigners(signers, isActivatedList);
+    }
+
+    // -------------------
+    // -- Get Functions --
+    // -------------------
+
+    // Checks if a signer is activated
+    function isSignerActivated(address _signer) external view override returns (bool isActivated) {
+        return _isActivatedSigner[_signer];
+    }
+
+    // Gets the closure status of an event
+    function getEvent(address tokenAddress, string calldata eventID)
+        external
+        view
+        override
+        returns (bool isEventClosed)
+    {
+        bytes32 eventIDHash = _hashString(eventID);
+        return !_isEventOngoing[tokenAddress][eventIDHash];
+    }
+
+    // Gets the claimed amount for a user in a specific event
+    function getClaimStatus(address tokenAddress, string calldata eventID, address userAddress)
+        external
+        view
+        override
+        returns (uint256 claimedAmount)
+    {
+        bytes32 eventIDHash = _hashString(eventID);
+        return _userClaimedAmount[tokenAddress][eventIDHash][userAddress];
+    }
+
+    // Gets the list of all signers
+    function getSigners() external view returns (address[] memory) {
+        return _signerSet.values();
+    }
+
+    function getClaimHash(address tokenAddress, string calldata eventID, address userAddress, uint256 amount)
+        public
+        pure
+        returns (bytes32)
+    {
+        bytes32 eventIDHash = _hashString(eventID);
+        return keccak256(abi.encode(tokenAddress, eventIDHash, userAddress, amount));
+    }
+
+    // -------------------
+    // -- Set Functions --
+    // -------------------
+
+    // Updates signers and their activation status
+    function updateSigners(address[] memory signerList, bool[] memory isActivatedList) external override onlyOwner {
+        _updateSigners(signerList, isActivatedList);
+    }
+
+    // Creates a new event with the specified ID and token address
+    function createNewEvent(address tokenAddress, string calldata eventID) external override onlyOwner {
```

Comment:
Let's add an additional parameter `bool startEvent` to let the admin choose whether to start the event right away or start it later.